### PR TITLE
Update VariablePart to match off-chain code

### PR DESCRIPTION
**Pull request by nksazonov**
Relates to #499

### Changes

Update `VariablePart` struct to include:
- [x] `outcome` as `SingleAssetExit[]`
- [x] `turnNum` and `isFinal` flag 

Update function logic (majorly in `ForceMove.sol`) to support aforementioned changes:
- [x] support changing `outcome` as `SingleAssetExit[]`
- [x] support adding `turnNum` and `isFinal` flag 

---

#### Code quality

- [x] I have written clear commit messages
- [x] I have performed a self-review of my own code
- [ ] This change does not have an unduly wide scope
- [x] I have separated logic changes from refactor changes (formatting, renames, etc.)
- [x] I have commented my code wherever necessary (can be 0)
- [ ] I have added tests that prove my fix is effective or that my feature works, if necessary
- [x] I have added comprehensive [godoc](https://go.dev/blog/godoc) comments 

#### Project management

- [x] I have assigned myself to this PR
- [x] I have linked the appropriate github issue
- [ ] I have assigned this PR to the appropriate GitHub project
- [ ] I have assigned this PR to the appropriate GitHub Milestone

### CODE REVIEW

**Comment by nksazonov** [comment]
Have a few doubts about the following:
- should a function for returning the last element of `VariablePart[]` be added? (e.g. for usage [here](https://github.com/statechannels/go-nitro/pull/508/files#diff-483aad04b1f90d19c78f0fd86f7ebb7869c6c194797c3d0774ce9e80dff5a2b8R91) and [here](https://github.com/statechannels/go-nitro/pull/508/files#diff-483aad04b1f90d19c78f0fd86f7ebb7869c6c194797c3d0774ce9e80dff5a2b8R191)) . The benefit from it is clearer code (repetitions are in a separate function), but the drawback is slightly increased gas usage (by 100 gas).

File: nitro-protocol/contracts/NitroAdjudicator.sol [hunk 1]
```diff
@@ -13,34 +13,28 @@ contract NitroAdjudicator is ForceMove, MultiAssetHolder {
...
-            appData,
-            outcomeBytes,
+            latestVariablePart,
             numStates,
             whoSignedWhat,
             sigs
         );
 
-        transferAllAssets(channelId, outcomeBytes, bytes32(0));
+        transferAllAssets(channelId, Outcome.encodeExit(latestVariablePart.outcome), bytes32(0));
```
Comment (geoknee) [inline nitro-protocol/contracts/NitroAdjudicator.sol:37]: There are different options here, but to me it seems slightly nicer to:

* pass the `latestVariablePart.outcome`
* have `transferAllAssets` encode and hash it

(rather than encoding it here, passing the bytes, and having `transferAllAssets` hash it and decode it).

**Comment by geoknee** [comment]
> * update `_hashState` to be called with `SingleAssetExit[] outcome`

We should tackle this in a follow-up PR. When this one is merged, we can make a new issue to track that. 

> * should a function for returning the last element of `VariablePart[]` be added? (e.g. for usage [here](https://github.com/statechannels/go-nitro/pull/508/files#diff-483aad04b1f90d19c78f0fd86f7ebb7869c6c194797c3d0774ce9e80dff5a2b8R91) and [here](https://github.com/statechannels/go-nitro/pull/508/files#diff-483aad04b1f90d19c78f0fd86f7ebb7869c6c194797c3d0774ce9e80dff5a2b8R191)) . The benefit from it is clearer code (repetitions are in a separate function), but the drawback is slightly increased gas usage (by 100 gas).

This could also be tackled in a follow-up PR. The links you gave don't seem to point to a particular line, so I'm not sure exactly where you propose to use such a function. In general I would say it is worth a little bit of gas to improve the quality of the code. 

This is reminding me that we should reintroduce some tracking of gas costs in CI, to make these decisions easier to make.

**Comment by geoknee** [comment]
@nksazonov this looks great to me! We need to do a bit more work before merging, however:

1. We need to get the CI tests to run on this branch. If you're comfortable, I would suggest rebasing this branch on `main` (which now includes #487). Then we should get a second github action running. 
2. I expect that to fail since this PR changes the contract API and we didn't update the tests yet. I'm afraid it will be a bit laborious :-/ . Happy to help out with that work, though.

File: nitro-protocol/contracts/interfaces/IForceMove.sol [hunk 2]
```diff
@@ -35,78 +35,64 @@ interface IForceMove {
...
+     * @param variableParts An ordered array of structs, each decribing the properties of the state channel that may change with each state update. Length is from 1 to the number of participants (inclusive).
+     * @param sigs An array of signatures that support the state with the `largestTurnNum`. There must be one for each participant, e.g.: [sig-from-p0, sig-from-p1, ...]
      * @param whoSignedWhat An array denoting which participant has signed which state: `participant[i]` signed the state with index `whoSignedWhat[i]`.
      * @param challengerSig The signature of a participant on the keccak256 of the abi.encode of (supportedStateHash, 'forceMove').
      */
     function challenge(
-        FixedPart calldata fixedPart,
-        uint48 largestTurnNum,
-        IForceMoveApp.VariablePart[] calldata variableParts,
-        uint8 isFinalCount, // how many of the states are final
```
Comment (geoknee) [inline nitro-protocol/contracts/interfaces/IForceMove.sol:49]: Glad to see this go!

File: nitro-protocol/contracts/interfaces/IForceMove.sol [hunk 3]
```diff
@@ -35,78 +35,64 @@ interface IForceMove {
...
     function conclude(
-        uint48 largestTurnNum,
-        FixedPart calldata fixedPart,
-        bytes memory appData,
-        bytes memory outcome,
+        FixedPart memory fixedPart,
+        IForceMoveApp.VariablePart memory latestVariablePart,
         uint8 numStates,
-        uint8[] calldata whoSignedWhat,
-        Signature[] calldata sigs
```
[the remark this answers is not in the document]
Comment (nksazonov) [reply nitro-protocol/contracts/interfaces/IForceMove.sol:109]: Sure

**Comment by nksazonov** [comment]
> > update _hashState to be called with SingleAssetExit[] outcome

> We should tackle this in a follow-up PR. When this one is merged, we can make a new issue to track that.

Has not noticed it on time. Let me revert these changes and force-push.

File: nitro-protocol/contracts/NitroAdjudicator.sol (hunk 1 above)
Comment (nksazonov) [reply nitro-protocol/contracts/NitroAdjudicator.sol:37]: This way looks better to me too.

File: nitro-protocol/contracts/interfaces/IForceMove.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (nksazonov) [reply nitro-protocol/contracts/interfaces/IForceMove.sol:109]: Adding `calldata` to `FixedPart` or `IForceMoveApp.VariablePart` and compilation results in `UnimplementedFeatureError:`. This is pretty strange, because `calldata` is present in current implementation of on-chain statechannels part in monorepo and compiling it finishes successfully.

**Comment by nksazonov** [comment]
> 2. I expect that to fail since this PR changes the contract API and we didn't update the tests yet. I'm afraid it will be a bit laborious :-/ . Happy to help out with that work, though.

I will start working on changing tests today then.

File: nitro-protocol/contracts/interfaces/IForceMove.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (geoknee) [reply nitro-protocol/contracts/interfaces/IForceMove.sol]: I can't remember, possibly! Happy to just go with `memory` for now if it unblocks us. We can make an issue to circle back to it.

**Comment by nksazonov** [comment]
The only thing left now is to update `EmbeddedApp.sol` to have the same `VariablePart` struct and test related to it.

**Comment by geoknee** [comment]
> The only thing left now is to update `EmbeddedApp.sol` to have the same `VariablePart` struct and tests related to it.

I would be in favour of simply removing or disabling `EmbeddedApp.sol` and tests, for now, because:

* We can reinstate that code as a part of #496 
  * The `EmbeddedApp` could become the `VirtualChannelApp`
  * It might be easier to write from scratch, possibly following Andrew's example https://github.com/statechannels/go-nitro/pull/427#issuecomment-1086025068
* It would be good to get the changes on this PR merged quickly

**Comment by nksazonov** [comment]
> I would be in favour of simply removing or disabling EmbeddedApp.sol and tests, for now

Okay, then I will just tune the tests for `EmbeddedApp.sol`.

**Comment by nksazonov** [comment]
@geoknee all tests are now passing.

File: nitro-protocol/test/contracts/ForceMove/challenge.test.ts [hunk 4]
```diff
@@ -169,6 +170,9 @@ describe('challenge', () => {
       const variableParts = states.map(state => getVariablePart(state));
       const fixedPart = getFixedPart(states[0]);
       const channelId = getChannelId(fixedPart);
+      console.log("variableParts");
+      console.log(variableParts);
```
Comment (geoknee) [inline nitro-protocol/test/contracts/ForceMove/challenge.test.ts:174]: ✂️

File: nitro-protocol/contracts/examples/HashLockedSwap.sol [hunk 5]
```diff
@@ -17,11 +17,10 @@ contract HashLockedSwap is IForceMoveApp {
     function validTransition(
         VariablePart memory a,
         VariablePart memory b,
-        uint48 turnNumB,
         uint256
     ) public override pure returns (bool) {
         // is this the first and only swap?
-        require(turnNumB == 4, 'turnNumB != 4');
+        require(b.turnNum == 4, 'turnNumB != 4');
```
Comment (geoknee) [inline nitro-protocol/contracts/examples/HashLockedSwap.sol:23]: Nit: 
```suggestion
        require(b.turnNum == 4, 'b.turnNum != 4');
```

File: nitro-protocol/src/contract/state.ts [hunk 6]
```diff
@@ -53,7 +55,12 @@ export interface VariablePart {
  * @returns the VariablePart, which usually changes during state channel updates
  */
 export function getVariablePart(state: State): VariablePart {
-  return {outcome: encodeOutcome(state.outcome), appData: encodeAppData(state.appData)};
+  return {
+    outcome: state.outcome,
+    appData: encodeAppData(state.appData),
```
Comment (geoknee) [inline nitro-protocol/src/contract/state.ts:60]: Outside the scope of this PR -- This seems a little strange, to me. I wouldn't expect we need to re-encode `appData` when extracting it from the state?

File: nitro-protocol/test/contracts/ForceMove/challenge.test.ts [hunk 7]
```diff
@@ -44,7 +45,7 @@ const chainId = process.env.CHAIN_NETWORK_ID;
 const participants = ['', '', ''];
 const wallets = new Array(3);
 const challengeDuration = 86400; // 1 day
-const outcome: Outcome = [{allocations: [], asset: Wallet.createRandom().address, metadata: '0x'}];
+const outcome: Outcome = [{allocations: [{destination: "0x00000000000000000000000096f7123e3a80c9813ef50213aded0e4511cb820f", amount: "0x01", allocationType: 1, metadata: "0x"}], asset: Wallet.createRandom().address, metadata: '0x'}];
```
Comment (geoknee) [inline nitro-protocol/test/contracts/ForceMove/challenge.test.ts:48]: Was this change necessary?

File: nitro-protocol/contracts/ForceMove.sol [hunk 8]
```diff
@@ -134,22 +127,22 @@ contract ForceMove is IForceMove, StatusManager {
...
+            Outcome.encodeExit(variablePartAB[1].outcome),
             turnNumRecord + 1,
-            isFinalAB[1]
+            variablePartAB[1].isFinal
         );
 
         // checks
 
-        bytes32 challengeOutcomeHash = keccak256(variablePartAB[0].outcome);
+        bytes32 challengeOutcomeHash = keccak256(Outcome.encodeExit(variablePartAB[0].outcome));
```
Comment (geoknee) [inline nitro-protocol/contracts/ForceMove.sol:145]: Outside the scope of this PR: might be nice to have a `hashOutcome` helper?

**Review by geoknee — APPROVED** [review]
👏

File: nitro-protocol/test/contracts/ForceMove/challenge.test.ts (hunk 7 above)
Comment (nksazonov) [reply nitro-protocol/test/contracts/ForceMove/challenge.test.ts]: It was done for testing purposes. Removed in [`0fea82b`](https://github.com/statechannels/go-nitro/pull/508/commits/0fea82b1634e81ba892dad33ad304ae4d36db885)

File: nitro-protocol/contracts/examples/HashLockedSwap.sol (hunk 5 above)
Comment (nksazonov) [reply nitro-protocol/contracts/examples/HashLockedSwap.sol]: Implemented in [`894e94d`](https://github.com/statechannels/go-nitro/pull/508/commits/894e94db7ea9d87ae2ce9118feb144561abe920c)

File: nitro-protocol/test/contracts/ForceMove/challenge.test.ts (hunk 4 above)
Comment (nksazonov) [reply nitro-protocol/test/contracts/ForceMove/challenge.test.ts]: Removed in [`0fea82b`](https://github.com/statechannels/go-nitro/pull/508/commits/0fea82b1634e81ba892dad33ad304ae4d36db885)

File: nitro-protocol/contracts/ForceMove.sol (hunk 8 above)
Comment (nksazonov) [reply nitro-protocol/contracts/ForceMove.sol:145]: Good idea.

File: nitro-protocol/src/contract/state.ts (hunk 6 above)
Comment (nksazonov) [reply nitro-protocol/src/contract/state.ts:60]: If I remember correctly, this was added to be correspond to `go-nitro` state encoding.
